sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs

```diff
@@ -0,0 +1,352 @@
+
+                readIndividualBuffer = await ReadLoopInternal(
+                    stream,
+                    buffer,
+                    offset: offset,
+                    minCountIndividualBuffer,
+                    maxCountIndividualBuffer,
+                    async,
+                    cancellationToken).ConfigureAwait(false);
+                if (readIndividualBuffer == 0 && newbuffer) // nothing was placed in a brand new array
+                {
+                    arrayPool.Return(buffer);
+                }
+                else if (newbuffer) // brand new array and we did place data in it
+                {
+                    streamPartition.BufferSet.Add(new BufferPartition
+                    {
+                        Buffer = buffer,
+                        DataLength = readIndividualBuffer
+                    });
+                }
+                else // added to an existing array that was not entirely filled
+                {
+                    latestBuffer.DataLength += readIndividualBuffer;
+                }
+
+                totalRead += readIndividualBuffer;
+
+            /* If we filled the buffer this loop, then quitting on min count is pointless. The point of quitting
+             * on min count is when the source stream doesn't have available bytes and we've reached an amount worth
+             * sending instead of blocking on. If we filled the available array, we don't actually know whether more
+             * data is available yet, as we limited our read for reasons outside the stream state. We should therefore
+             * try another read regardless of whether we hit min count.
+             */
+            } while (
+                readIndividualBuffer != 0 &&                                                 // stream is done if this value is zero; no other check matters
+                totalRead < maxCount &&                                                      // stop filling the partition if we've hit the max size of the partition
+                (totalRead < minCount || readIndividualBuffer == maxCountIndividualBuffer)); // stop filling the partition if we've reached min count and we know we've hit at least a pause in the stream
+
+            return streamPartition;
+        }
+
+        /// <summary>
+        /// Loops Read() calls into buffer until minCount is reached or stream returns 0.
+        /// </summary>
+        /// <returns>Bytes read.</returns>
+        /// <remarks>
+        /// This method may have read bytes even if it has reached the confirmed end of stream. You will have to call
+        /// this method again and read zero bytes to get that confirmation.
+        /// </remarks>
+        private static async Task<int> ReadLoopInternal(Stream stream, byte[] buffer, int offset, int minCount, int maxCount, bool async, CancellationToken cancellationToken)
+        {
+            if (minCount > maxCount) // nonsensical; don't allow caller methods to produce this
```

Comment:
Another super-nit comment thing - let's remove ```nonsensical``` and ```useless``` in this and the following comment.